blog: using kafka with node-red (#1758)
FlowFuse/website: src/blog/2024/03/using_kafka_with_node-red.md

<!--more-->

To address these challenges, various manufacturing and automobile plants have embraced technologies like Apache Kafka. Kafka provides a distributed streaming platform that enables the efficient handling of real-time data streams. By leveraging Kafka, we can aggregate, process, and analyze data in real-time seamlessly. However, despite its extensive benefits, Kafka also has some limitations. But there's no need to worry. Throughout this comprehensive guide, we will delve into everything about Kafka, including its pros and cons, its applications in the real world, and how to use it efficiently with Node-RED applications.

## What exactly is Kafka?
Apache Kafka is a platform for distributed data streaming that allows for the publishing, subscribing, storing, and processing of streams of records in real-time. It is intended to handle data streams from multiple sources and to deliver them to multiple consumers. In essence, it can move large quantities of data in real-time from any source to any destination, simultaneously.

> 

Reviewer: @sumitshinde-84 Can we pull out the "What is Kafka" into a different post? We'd have two posts, one high level. And one low level and we cross link

> 

Author: @ZJvandeWeg That sounds great. I think we should create two parts for this. In Part 1, as you suggested, we will cover the high-level overview of Kafka, including what Kafka is, its core components, applications, and limitations. In Part 2, we will explain how we can use it with Node-RED.

---

> title: Using Kafka with Node-Red - Part 1
> subtitle: Getting started with Kafka.
> description: Getting started with Kafka.
> date: 2024-03-26
> authors: ["sumit-shinde"]
> image: /blog/2024/03/images/using-kafka-with-node-red.png

Reviewer: ```suggestion
title: How Kafka is applied in manufacturing
subtitle: An overview of Kafka -- How it's applied for industrial applications, and how it works
description: An overview of Kafka -- How it's applied for industrial applications, and how it works
date: 2024-03-15
authors: ["sumit-shinde"]
image: /blog/2024/03/images/using-kafka-with-node-red.png
```

Tile would need updating

Reviewer 2: I second this.  There isn't much discussion of Node-RED in the article, which is fine.  Just don't want to mislead people.

Reviewer 2: @Yndira-E Can you update the image to match this new title please?

tags:
    - posts

>     - node-red
> ---

Reviewer: ```suggestion
    - node-red
    - kafka
---
```

**2. Producers:**
- Producers: Producers are like architects of data flow. They decide where to send records within a topic. This decision can be balanced using a round-robin or directed by a record key for specific purposes, such as maintaining order.

**3. Brokers:**
- Definition: Brokers are the backbone servers in a Kafka cluster.
- Tasks: Brokers store data, handle requests from both producers and consumers, and maintain the integrity and persistence of data. They also manage the critical task of tracking offsets, which determine the position of consumers within partitions.

**4. Consumers and Consumer Groups:**
- Consumers: These entities read data from brokers. They subscribe to one or more topics and pull data from the specific partitions they are interested in.
- Consumer Groups: Consumers collaborate in groups to scale data processing. Kafka dynamically assigns each consumer in a group a set of partitions from the subscribed topics, ensuring that each partition is processed by only one consumer within the group.

**5. Offsets**
- Definition: Offsets act as unique identifiers for records within a partition. They denote the position of a consumer in the partition.

> - Function: As consumers read records, they increment their offset. This allows them to resume processing from where they left off, crucial for handling failures or restarts. Kafka stores offset information in a specialized topic for easy recovery.

Reviewer 2: ```suggestion
- Function: As consumers read records, they increment their offset. This allows them to resume processing from where they left off, which is crucial for handling failures or restarts. Kafka stores offset information in a specialized topic for easy recovery.
```

---
title: How Kafka is applied in manufacturing
subtitle: An overview of Kafka -- How it's applied for industrial applications, and how it works
description: An overview of Kafka -- How it's applied for industrial applications, and how it works
date: 2024-03-15
authors: ["sumit-shinde"]
image: /blog/2024/03/images/using-kafka-with-node-red.png
tags:
    - posts
    - node-red
    - kafka
---

> Have you ever wondered how manufacturing and automotive industries can effectively manage the vast amount of real-time data generated by sensors and systems throughout the production process? A few years back, these industries faced major obstacles in handling the large volume of real-time data produced by sensors placed across the production line. Even nowadays many industries continue to grapple with similar challenges. Traditional data management systems struggle to process and analyze this data in real-time, leading to inefficiencies in operational activities and decision-making. To address these challenges, various manufacturing and automobile plants have embraced technologies like Apache Kafka.

Reviewer 3: ```suggestion
Have you ever wondered how manufacturing and automotive industries can effectively manage the vast amount of real-time data generated by sensors and systems throughout the production process? A few years back, these industries faced major obstacles in handling the large volume of real-time data produced by sensors placed across the production line. Even today many industries continue to grapple with similar challenges. Traditional data management systems struggle to process and analyze this data in real-time, leading to inefficiencies in operational activities and decision-making. To address these challenges, various manufacturing and automobile plants have embraced technologies like Apache Kafka.
```

<!--more-->

Kafka provides a distributed streaming platform that enables the efficient handling of real-time data streams. By leveraging Kafka, we can aggregate, process, and analyze data in real-time seamlessly. This guide provides a high-level overview of Kafka, covering its definition, components, functionality, applications, and limitations.

> ## What exactly is Kafka?

Reviewer 3: ```suggestion
## What is Kafka?
```
